### ISSUE TITLE
API question: cursors and databases

### ISSUE THREAD

**Issue by os12** [pmwkaa/sophia#154]
Hi Dmitry, I've started stated testing Sophia and hit an issue: a cursor-based walk returns keys across databases. Assume two databases, "db1" and "db2". One of them is opened and keys are fetched using a cursor:

```
    void *env = sp_env();

    sp_setstring(env, "sophia.path", ".store", 0);
    sp_setstring(env, "db", db_name.c_str(), 0);
    if (sp_open(env) == -1)
        return FromStringError(LastSophiaError(env));

    void *db = sp_getobject(env, ("db." + db_name).c_str());
    if (db == nullptr)
        return FromStringError(LastSophiaError(env));

    void *cursor = sp_cursor(env);
    void *doc = sp_document(db);
    sp_setstring(doc, "order", ">=", 0);

    while ((doc = sp_get(cursor, doc))) {
        int size;
        auto key_ptr = static_cast<const char *>(sp_getstring(doc, "key", &size));
        ...
        ...
    }
```

Am I missing something here? Do cursors go across databases? If so, how do I provide an additional filter to limit iteration to a single DB?

Thanks a lot!
Oleg.

**Comment by pmwkaa** [comment]
Hi, not quite sure what the issue in here. There is no way records can overlap this way, only if you use different db pointers during sp_document() creation. Are you sure it sets correct name for sp_getobject()?

sophia.path specifies main directory, where all databases will be stored by name. For example, db.mydb will be placed in sophia.path/mydb

**Comment by os12** [comment]
Right, both the cursor-based "read" path and the transaction-based "write" path start with the same code (the first fragment). I do see per-database subdirectories created inside `.store` as well as `log` (which seems to contain the last few mutations).

I know you have tests that cover multi-db mutations. Yet, do you have a multi-db cursor test?

**Comment by pmwkaa** [comment]
Could you please specify your work case you are trying to achieve? One cursor is designed to iterate only one database at a time. If you need to make a join between databases, you need to open a second cursor, and so on.

**Comment by os12** [comment]
> Could you please specify your work case you are trying to achieve? One cursor is designed to iterate only one database at a time

Yes, that's what I expected. Yet the cursor returns keys from an adjacent DB, which is very surprising.

**Comment by os12** [comment]
Please see #156 for the repro - a cursor-based walk in DB2 should find no keys (yet it fetches everything from the DB1).

**Comment by pmwkaa** [comment]
Now i understand what you meant. This is known issue, which might be quite confusing. Sophia does not have persistent catalog of created databases, they must be predefined each time before `sp_open()`. This is done to support easier Sophia integration with DBMS's which have its own database catalog.

Open procedure will find and recover every specified database directory by its name. After that it will reply write-ahead log records. To identify database in write-ahead log `db.name.id` is used: http://sophia.systems/v2.2/conf/db.html.  If database is not specified it will be set in sequential order of database definition.

**Comment by os12** [comment]
Umm... I don't follow. Is there a way for me do deal with this issue? 
* My application needs to read/write distinct DBs at random times (otherwise, I'd have to impose prefixes onto record names to fake that)
* The "read" is usually started by opening a DB by name and then listing records using a cursor (they must be ordered).

If this does not work, I will have to create distinct directories to isolate the DBs... Are there downsides?

**Comment by pmwkaa** [comment]
Yes, you just need to set unique `db.name.id` for each database. This will allow to properly distinct databases during open and log reply.

**Comment by os12** [comment]
OK, I've extended the code to set `db.db1.id` bit it's failing with this:
```
sophia/runtime/sr_conf.c:100 db.db1.id is read-only
```

Could you take a look at my second commit in #156 please? What am I missing?

**Comment by os12** [comment]
Also, let me use this comment to collect misc notes/comments:
1. The list of options at http://sophia.systems/v2.2/conf/db.html is hard to understand as it mentioned things like `db.name.id`
   * it should really read something like `db.$(db_name).id` as the string is meaningless otherwise
1. The across-db nature of the log file makes things very tricky:
   * it should be documented somewhere in **BOLD**
   * it's hard to manage to these IDs in the environments that create databases on the fly
      * basically, you are forcing the caller to keep a "db_name to id" map and then provide this extra metadata at every `sp_open()` call
    * `sp_open()` is expensive and so the DBs should not be closed

**Comment by pmwkaa** [comment]
My fault, sorry about that. I completely forgot that custom id generation scheme been removed in v2.2 (so recommended scheme was actual only for previous version). Right now, the only way to make this work is to always define same databases in same order. I was planning to completely remake it.

It is also possible to force compaction and wait for its completion, this will guarantee that write-ahead records will not have database references. This will also increase open time dramatically.

Could you describe your use-case in more details? i might have some thoughts how to make it in optimal way.
